cli/pkg/workspace/plugin/kubeconfig.go

```diff
@@ -132,187 +134,189 @@ func (o *UseWorkspaceOptions) Run(ctx context.Context) error {
-		newServerHost = u.String()
+	// validate early for valid name
+	err = o.validate(ctx)
+	if err != nil {
+		return err
+	}
 
-	case "":
+	switch {
+	// o.Name empty and home workspace should be always one after another for fallthrough to
+	// print error and default to home.
+	case o.Name == "":
 		defer func() {
 			if err == nil {
 				_, err = fmt.Fprintf(o.Out, "Note: 'kubectl ws' now matches 'cd' semantics: go to home workspace. 'kubectl ws -' to go back. 'kubectl ws .' to print current workspace.\n")
 			}
 		}()
 		fallthrough
 
-	case "~", home:
-		homeWorkspace, err := o.kcpClusterClient.Cluster(core.RootCluster.Path()).TenancyV1alpha1().Workspaces().Get(ctx, "~", metav1.GetOptions{})
+	case o.Name == "~" || o.Name == home:
+		newServerHost, err := o.moveHome(ctx)
 		if err != nil {
 			return err
 		}
-		config, err := o.ClientConfig.ClientConfig()
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
+
+		// move to 'my' root
+	case o.Name == ":":
+		newServerHost, err := o.moveCurrentRoot(ctx)
 		if err != nil {
 			return err
 		}
-		u, _, err := pluginhelpers.ParseClusterURL(config.Host)
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
+
+	case o.Name == ".":
+		cfg, err := o.ClientConfig.ClientConfig()
 		if err != nil {
-			return fmt.Errorf("current URL %q does not point to a workspace", config.Host)
+			return err
 		}
-		uh, err := url.Parse(homeWorkspace.Spec.URL)
+		return printCurrentWorkspace(o.Out, cfg.Host, shortWorkspaceOutput(o.ShortWorkspaceOutput), nil)
+
+	case o.Name == "-":
+		newServerHost, err := o.movePrevious(ctx, currentContext)
 		if err != nil {
-			return fmt.Errorf("invalid home workspace URL %q: %w", homeWorkspace.Spec.URL, err)
+			return err
 		}
+		return printCurrentWorkspace(o.Out, newServerHost, shortWorkspaceOutput(o.ShortWorkspaceOutput), nil)
 
-		// We keep the old host, and append homeworkspace url. This allows users to have front-proxy
-		// host to be different from the workspace host.
-		newServerHost = u.Scheme + "://" + path.Join(u.Host, uh.Path)
-
-	case ".":
-		cfg, err := o.ClientConfig.ClientConfig()
+	case strings.Contains(o.Name, ".."):
+		newServerHost, err := o.moveBackwards(ctx)
 		if err != nil {
 			return err
 		}
-		return currentWorkspace(o.Out, cfg.Host, shortWorkspaceOutput(o.ShortWorkspaceOutput), nil)
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
 
-	default:
-		cluster := logicalcluster.NewPath(o.Name)
-		if !cluster.IsValid() {
-			return fmt.Errorf("invalid workspace name format: %s", o.Name)
-		}
-		config, err := o.ClientConfig.ClientConfig()
+		// root legacy move
+	case o.Name == core.RootCluster.String():
+		newServerHost, err := o.moveRootLegacy(ctx)
 		if err != nil {
 			return err
 		}
-		u, currentClusterName, err := pluginhelpers.ParseClusterURL(config.Host)
+		return o.commitConfig(ctx, currentContext, newServerHost, nil)
+
+	case strings.HasPrefix(o.Name, "system") || strings.HasPrefix(o.Name, ":system"):
```

Comment:
why is `system:` different from `root:`?

Reply:
I think we can remove this and just use `:system:admin` and will just be handled as "absolute any root"

Reply:
Added comments. I think it was originally a separate case due to permissions too. Merged it to generic code but it got a complex mess again :/ better have explisting check than if/else/if/else in other places :/